### ISSUE TITLE
[Bug] Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal() throws a null reference exception

### ISSUE THREAD

**Issue by jimitndiaye** [AzureAD/microsoft-identity-web#1624]
**Which version of Microsoft Identity Web are you using?**
Note that to get help, you need to run the latest version. 

Version is 1.22.3

**Where is the issue?**
* Web app
    * [ ] Sign-in users
    * [x] Sign-in users and call web APIs
* Web API
    * [ ] Protected web APIs (validating tokens)
    * [ ] Protected web APIs (validating scopes)
    * [ ] Protected web APIs call downstream web APIs
* Token cache serialization
     * [ ] In-memory caches
     * [ ] Session caches
     * [ ] Distributed caches
* Other (please describe)

**Is this a new or an existing app?**

This is a new app
**Repro**
I have the exact same problem as in #1507 . There's the setup code:
```csharp
            services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
                .AddMicrosoftIdentityWebApp(
                    openIdConnectScheme: OpenIdConnectDefaults.AuthenticationScheme,
                    displayName: "Azure AD B2C",
                    configureMicrosoftIdentityOptions: options =>
                        ConfigureMicrosoftIdentityOptions(
                            options,
                            shellSettings),
                    configureCookieAuthenticationOptions: ConfigureCookieAuthenticationOptions,
                    cookieScheme: null,
                    subscribeToOpenIdConnectMiddlewareDiagnosticsEvents: env.IsDevelopment())
                .EnableTokenAcquisitionToCallDownstreamApi()
                .AddDownstreamWebApi(AuthenticationConstants.IdentityApiName,
                    options => { Configuration.GetSection("Portal:Authentication:Api").Bind(options); })
                .AddDistributedTokenCaches();
```
```ConfigureMicrosoftIdentityOptions``` is defined below:
```csharp
    protected virtual void ConfigureMicrosoftIdentityOptions(
        MicrosoftIdentityOptions options,
        ShellSettings shellSettings)
    {
        Configuration.GetSection(PortalAuthenticationConfigSection).Bind(options);
        var signInPolicies = Configuration.GetSection(PortalAuthenticationConfigSection).Get<SignInPolicyOptions>();
        options.SignUpSignInPolicyId = GetSignInPolicy(signInPolicies);
        var urlPrefix = !string.IsNullOrEmpty(shellSettings.RequestUrlPrefix)
            ? shellSettings.RequestUrlPrefix.Trim('/').PadLeft(1, '/')
            : null;
        options.CallbackPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.CallbackPath}";
        options.SignedOutCallbackPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.SignedOutCallbackPath}";
        options.SignedOutRedirectUri = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.SignedOutPath}";
        options.RemoteSignOutPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.RemoteSignOutPath}";
        options.ErrorPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.ErrorPath}";
        options.ResetPasswordPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.ResetPasswordPath}";
        options.AccessDeniedPath = $"{urlPrefix}/{AuthenticationConstants.CallbackPaths.AccessDeniedPath}";
        options.GetClaimsFromUserInfoEndpoint = true;
        //options.SaveTokens = true;
        options.SignInScheme = IdentityConstants.ExternalScheme;
        //options.ForwardDefault = IdentityConstants.ExternalScheme;
        options.TokenValidationParameters = new()
        {
            NameClaimType = "name"
        };
}

```

I then inject an instance of IDownstreamWebApi into a controller and call it as follows:
```csharp
                    var organizations =  await _api.CallWebApiForUserAsync<OrganizationSummary[]>(AuthenticationConstants.IdentityApiName,
                        api =>
                        {
                            api.RelativePath = "organizations/oauth2";
                            api.HttpMethod = HttpMethod.Get;
                        });
```
I can sign in fine using the above code, so authentication works ok. It's when I need to call the downstream api that I run into issues
**Expected behavior**
A clear and concise description of what you expected to happen (or code).
I should be able to call the downstream API without errors
**Actual behavior**
IDownsteamWebApi.CallWebApiForUserAsync throws NullReferenceException with the following stack trace:
```
   at Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal()
   at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.<GetAuthenticationResultForUserAsync>d__16.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.ConfiguredTaskAwaitable`1.ConfiguredTaskAwaiter.GetResult()
   at Microsoft.Identity.Web.DownstreamWebApi.<CallWebApiForUserAsync>d__5.MoveNext()
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at MyNamespace.OnboardingController.<Index>d__19.MoveNext() 
```

**Possible solution**

**Additional context / logs / screenshots / link to code**
I tried manually configurating the ConfidentialClientApplicationOptions but it made no difference.
```csharp
            services.PostConfigure<ConfidentialClientApplicationOptions>(OpenIdConnectDefaults.AuthenticationScheme,
                options =>
                {
                    Configuration.GetSection(PortalAuthenticationConfigSection).Bind(options);
                    options.EnablePiiLogging = env.IsDevelopment();
                    Debug.Assert(!string.IsNullOrEmpty(options.Instance), "!string.IsNullOrEmpty(options.Instance)");
                });
```
The above code runs ok and i can confirm that Instance was not null at that point but it some how still gets overwritten somewhere

Add any other context about the problem here, such as logs and screenshots, or even links to code.

**Comment by jimitndiaye** [comment]
Upon further investigation it seems that might be in Microsoft.Identity.Web.TokenAcquisition
```
        public string GetEffectiveAuthenticationScheme(string? authenticationScheme)
        {
            if (authenticationScheme != null)
            {
                return authenticationScheme;
            }
            else
            {
                return _serviceProvider.GetService<IAuthenticationSchemeProvider>()?.GetDefaultAuthenticateSchemeAsync()?.Result?.Name ??
                    ((CurrentHttpContext?.GetTokenUsedToCallWebAPI() != null)
                    ? JwtBearerDefaults.AuthenticationScheme : OpenIdConnectDefaults.AuthenticationScheme);
            }
        }
```
For what ever reason when the authentication scheme is not explicitly specified it might not be returning the right scheme. When I explicitly set the authentication scheme to OpenIdConnectionDefaults.AuthenticationScheme it doesn't throw a null reference exception:
```csharp
                    var organizations = await _api.CallWebApiForUserAsync<OrganizationSummary[]>(
                        AuthenticationConstants.IdentityApiName,
                        api =>
                        {
                            api.RelativePath = "organizations/oauth2";
                            api.HttpMethod = HttpMethod.Get;
                        },
                        authenticationScheme: OpenIdConnectDefaults.AuthenticationScheme);
```

**Comment by thuatnguyenvan** [comment]
Actually, this is a bug of Microsoft.Identity.Web from version 1.16.1 to latest 1.22.1.

.NET 6
Microsoft.Identity.Web from version 1.16.1 to 1.22.1 : not working
Microsoft.Identity.Web from version < 1.16.1 : OK

**Comment by jmprieur** [comment]
@thuatnguyenvan what do you have in your authorize attribute on the controller or the controller action? do you specify an auth scheme?

**Comment by thuatnguyenvan** [comment]
@jmprieur [jmprieur](https://github.com/jmprieur) My problem is from `DefaultScheme = "My_Cokie_Name"` instead of `DefaultScheme = OpenIdConnectDefaults.AuthenticationScheme` . I guest it was throwing an exception in [Microsoft.AspNetCore.Authentication.Abstractions](https://github.com/aspnet/HttpAbstractions/tree/master/src/Microsoft.AspNetCore.Authentication.Abstractions)/AuthenticationOptions.cs

```
services.AddAuthentication(opt =>
            {
                /* - opt.DefaultScheme = "My_Cokie_Name" will not working from Microsoft.Identity.Web from 
                 * version 1.16.1 to 1.22.1 and .NET 6
                 * - To fix: 
                 *      opt.DefaultScheme = OpenIdConnectDefaults.AuthenticationScheme
                 */
                opt.DefaultScheme = OpenIdConnectDefaults.AuthenticationScheme;
                opt.DefaultChallengeScheme = OpenIdConnectDefaults.AuthenticationScheme;
                opt.DefaultSignInScheme = "My_Cokie_Name";
            })
            .AddMicrosoftIdentityWebApp(opts =>
            {
                .....
                opts.SignInScheme =  "My_Cokie_Name";
                opts.SignOutScheme =  "My_Cokie_Name";
                ...
             }, opt =>
            {
                opt.LoginPath = "your LoginPath".;
                opt.LogoutPath = "your LogoutPath";
                opt.Cookie.Name = "My_Cokie_Name";
                ...
             }, OpenIdConnectDefaults.AuthenticationScheme, "My_Cokie_Name")
              .EnableTokenAcquisitionToCallDownstreamApi(options =>
              {
                  configuration.Bind("Azure", options);
              }, new string[] { "openid" })
              .AddDistributedTokenCaches();
```

**Comment by jmprieur** [comment]
@jimitndiaye : did you try https://www.nuget.org/packages/Microsoft.Identity.Web/1.23.1?
we fixed a similar issue.

**Comment by froens** [comment]
@jmprieur 

I have the same problem. I tried to upgrade to 1.23.1, but this doesn't solve my issue.

Instance-property of MergedOptions is null, thus this [line](https://github.com/AzureAD/microsoft-identity-web/blob/c29ba1ccc9340a2c36d72a523dcde4a6b2b65613/src/Microsoft.Identity.Web/MergedOptions.cs#L378) is throwing an exception.

I noticed that the optionsmonitors cache-property is alway empty. But I'm not sure how this should have been populated (and if this is even relevant).

This is my setup:

```csharp
authBuilder.AddMicrosoftIdentityWebApp(opt =>
{
    opt.SignInScheme = IdentityConstants.ExternalScheme;
    opt.SignOutScheme = IdentityConstants.ExternalScheme;
    opt.TenantId = azureAdOptions.TenantId;
    opt.Instance = "https://login.microsoftonline.com";
    opt.Authority = $"https://login.microsoftonline.com/{azureAdOptions.TenantId}/v2.0";
    opt.ClientId = azureAdOptions.ClientId;

    opt.ResponseType = OpenIdConnectResponseType.CodeIdToken;
    opt.CallbackPath = "/signin-aad";
    opt.SaveTokens = true;
    opt.Scope.Add("email");

    opt.ClientCertificates = new List<CertificateDescription>()
    {
        CertificateDescription.FromCertificate(cert)
    };
}, openIdConnectScheme: "aad", displayName: "Azure AD", cookieScheme: null)
.EnableTokenAcquisitionToCallDownstreamApi()
.AddInMemoryTokenCaches()
.AddMicrosoftGraph(x =>
{
    var scopes = new[] { "User.Read" };

    // single-tenant apps must use the tenant ID from the Azure portal
    var tenantId = azureAdOptions.TenantId;

    // Values from app registration
    var clientId = azureAdOptions.ClientId;
    var clientSecret = azureAdOptions.ClientSecret;

    //This seems a bit odd - I don't think we should build the SP
    var sp = services.BuildServiceProvider();
    var ctx = sp.GetRequiredService<IHttpContextAccessor>();
    var ta = sp.GetRequiredService<ITokenAcquisition>();

    var oboToken = ta.GetAccessTokenForUserAsync(scopes, authenticationScheme: "aad", user: ctx.HttpContext.User).Result;

    var cca = ConfidentialClientApplicationBuilder
        .Create(clientId)
        .WithTenantId(tenantId)
        .WithClientSecret("<REDACTED>")
        .Build();

    // DelegateAuthenticationProvider is a simple auth provider implementation
    // that allows you to define an async function to retrieve a token
    // Alternatively, you can create a class that implements IAuthenticationProvider
    // for more complex scenarios
    var authProvider = new DelegateAuthenticationProvider(async (request) => {
        // Use Microsoft.Identity.Client to retrieve token
        var assertion = new UserAssertion(oboToken);
        var result = await cca.AcquireTokenOnBehalfOf(scopes, assertion).ExecuteAsync();

        request.Headers.Authorization =
            new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", result.AccessToken);
    });

    return new GraphServiceClient(authProvider);
}, new string[] { "user.read" })
.AddInMemoryTokenCaches();
```

**Comment by kyledevans** [comment]
I am encountering the same issue where it throws a null reference exception.  I'm trying to create a SPA application that communicates to my backend service API (the backend for frontend model).  Ideally this application would utilize a SSO solution like Azure AD, but often times for development, testing, and debugging purposes I find that it's easier to use a solution that implements an internally managed username/password store.

I am utilizing the Asp.Net Core Identity framework with an Entity Framework store in SQL Server to persist usernames/passwords.  The frontend SPA application will then authenticate against an API, and then call protected endpoints with a simple JWT token.  I am currently trying to figure out how to implement middleware that will determine if the "Authorization: Bearer <...>" header has been issued by Azure AD, or by my own internal JWT generator.  From there it should delegate to the appropriate middleware.

So as a proof of concept, I've created a default "Selector" middleware, that for now just hardcodes to forward to the AzureAd scheme:
```
    authBuilder.AddPolicyScheme(JwtBearerDefaults.AuthenticationScheme, "Selector", options => {
        options.ForwardDefaultSelector = context => {
            return "AzureAd";
        };
    });
```
Azure AD is configured as such:
```
    authBuilder.AddMicrosoftIdentityWebApi(jwtOptions => {
        }, msOptions => {
            //msOptions.Auth
            builder.Configuration.GetSection("AzureAd").Bind(msOptions);
        }, "AzureAd")
        .EnableTokenAcquisitionToCallDownstreamApi(msIdentOptions => {
            builder.Configuration.GetSection("AzureAd").Bind(msIdentOptions);    // THIS NEVER RUNS!  WHY?
            Console.WriteLine("BALAHAHAHAHAHA");
        })
        .AddMicrosoftGraph(graphOptions => {
            builder.Configuration.GetSection("MicrosoftGraph").Bind(graphOptions);
        })
        .AddInMemoryTokenCaches();
```
I've posted the application code I'm developing, it's a bit in a state of disarray at the moment because I'm trying to overhaul how I perform user management.  [Here is the repository](https://github.com/kyledevans/kiddo).  I can try and reduce this down to a more minimal reproduction of the issue if requested, but that will take a bit of time since I haven't spent a lot of time learning the ins and outs of the identity framework.

**Comment by oledid** [comment]
Please see if this comment is relevant https://github.com/AzureAD/microsoft-identity-web/issues/1507#issuecomment-1076338507

**Comment by kyledevans** [comment]
@oledid That comment illuminated the perhaps oddly named parameter "jwtBearerScheme".  However I'm still having the same issue.

In my usecase, I am still calling the Graph API.  What I am trying to accomplish is API authentication against JWT tokens from multiple issuers (Azure AD and my own internal JWT generator).  So the calls to my API will have the header "Authorization: Bearer <JWT>".  Calls to Microsoft Graph from within my API will use the on behalf flow, and will only be performed when the JWT issuer is Azure AD.

So basically my SPA application will allow end users to choose whether to login with either an internal username/password, or utilize Azure AD for authentication.  Most of my API endpoints don't worry which authentication method is used, so I simply put the bare [Authorize] attribute on the actions.  To support this, I have a simple authorization scheme that will parse the JWT token, and then forward to either the Microsoft Identity Web framework for validation, or use my validator for internal username/password tokens.

I have to admit that I am pretty lost with a lot of this.  Between learning OAuth, AzureAd, and the intricacies of the Asp.Net Core Identity framework, I'm struggling quite a bit.

I made another attempt at configuring my Program.cs to support multiple JWT issuer tokens, but I still get the null reference exception.  Here's the updated version:
```
    var authBuilder = builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme);

    authBuilder.AddMicrosoftIdentityWebApi(builder.Configuration, configSectionName: "AzureAd", jwtBearerScheme: "AzureAd")
        .EnableTokenAcquisitionToCallDownstreamApi()
        .AddMicrosoftGraph(builder.Configuration.GetSection("MicrosoftGraph"))
        .AddInMemoryTokenCaches();

    authBuilder.AddPolicyScheme(JwtBearerDefaults.AuthenticationScheme, "Selector", options => {
        options.ForwardDefaultSelector = context => {
            return "AzureAd";
        };
    });
```

**Comment by oledid** [comment]
@kyledevans I see you are using `JwtBearerDefaults.AuthenticationScheme` here. This might be a trap you're stepping into. It's value is just a constant string "Bearer". It does nothing to enable Jwt, it's only there as an example name for the authentication scheme. If you use multiple authentication schemes it might be better to create your own names.

Implementation of JwtBearerDefaults:
```csharp
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

namespace Microsoft.AspNetCore.Authentication.JwtBearer
{
    /// <summary>
    /// Default values used by bearer authentication.
    /// </summary>
    public static class JwtBearerDefaults
    {
        /// <summary>
        /// Default value for AuthenticationScheme property in the JwtBearerAuthenticationOptions
        /// </summary>
        public const string AuthenticationScheme = "Bearer";
    }
}
```

Maybe explicitly name these things to make it easier to not mix up stuff

```csharp
public class MyAuthenticationSchemes
{
    public const string MyAuthentication = "MyAuthentication";
    public const string AuthenticateWithAzureAd = "AuthenticateWithAzureAd";

  // I usually use nameof for these things,
  // public const string MyConstant = nameof(MyConstant);
}
```

```csharp
builder.Services.AddAuthentication(MyAuthenticationSchemes.AuthenticateWithAzureAd)
  .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"))
  .EnableTokenAcquisitionToCallDownstreamApi()
  .AddMicrosoftGraph(builder.Configuration.GetSection("MicrosoftGraph"))
  .AddInMemoryTokenCaches();

builder.Services.AddAuthentication(MyAuthenticationSchemes.MyAuthentication)
  .Add// ... i don't know how to implement your authentication
```

It's one of the traps I have stepped into myself at least :)

**Comment by kyledevans** [comment]
@oledid I read your comments and tried your suggestions, but haven't been able to resolve the issue.

Here's the most recent version of my code:
```
    const string cSelector = "Selector", cAspNetIdentity = "AspNetIdentity", cAzureAd = "AzureAd";

    builder.Services
        .AddAuthentication(cSelector)
        .AddPolicyScheme(cSelector, cSelector, options => {
            options.ForwardDefaultSelector = context => {
                context.Request.Headers.TryGetValue(HeaderNames.Authorization, out var authHeader);
                if (authHeader.Count() == 0) return cAzureAd;
                var encodedToken = authHeader.First().Substring(JwtBearerDefaults.AuthenticationScheme.Length + 1);
                var jwtHandler = new JwtSecurityTokenHandler();
                var token = jwtHandler.ReadJwtToken(encodedToken);

                Log.Logger.Warning($"JWT Issuer: {token.Issuer}");

                if (token.Issuer.StartsWith("https://login.microsoftonline.com/")) return cAzureAd;
                else return cAspNetIdentity;
            };
        })
        .AddJwtBearer(cAspNetIdentity, options => {
            string? issuer = System.Reflection.Assembly.GetExecutingAssembly().GetName().Name;
            if (issuer == null) throw new Exception("Unable to determine Issuer for JWTs.");

            options.TokenValidationParameters = new() {
                ValidateIssuerSigningKey = true,
                IssuerSigningKey = new SymmetricSecurityKey(System.Text.Encoding.UTF8.GetBytes(builder.Configuration.GetRequiredSection("SpaConfiguration").GetValue<string>("AspNetSecret"))),
                ValidIssuer = issuer,
                ValidAudiences = new[] { $"{issuer}.Access", $"{issuer}.Refresh" },
                ClockSkew = TimeSpan.Zero
            };
        })
        .AddMicrosoftIdentityWebApi(builder.Configuration.GetSection("AzureAd"), cAzureAd)
        .EnableTokenAcquisitionToCallDownstreamApi()
        .AddMicrosoftGraph(builder.Configuration.GetSection("MicrosoftGraph"))
        .AddInMemoryTokenCaches();
```
I've tried many variations of this, including your edited comment where I make adding each scheme a new call.  However each time I end up with the same result.  I'm still getting the null reference exception.

I'm wondering if the microsoft-identity-web library isn't properly handling things like PolicySchemeHandlers (via AddPolicyScheme)?  Of course it could be a mistake on my part that I still need to work out.

**Comment by oledid** [comment]
@kyledevans where exactly are you getting the `Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal()` `NullReferenceException` now? what login do you use to trigger it and where does it throw?

**Comment by kyledevans** [comment]
Here's my stack trace:
```
Status Code: 0
Microsoft.Graph.ServiceException: Code: generalException
Message: An error occurred sending the request.

 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Microsoft.Identity.Web.MergedOptions.PrepareAuthorityInstanceForMsal()
   at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
   at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
   at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
   at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
   at Microsoft.Graph.UserRequest.GetAsync(CancellationToken cancellationToken)
   at Kiddo.Web.Models.ProfileModel.EnsureProfile() in C:\Users\Kyle\source\repos\Kiddo\Kiddo.Web\Models\ProfileModel.cs:line 83
   at Kiddo.Web.Controllers.ProfileController.EnsureProfile() in C:\Users\Kyle\source\repos\Kiddo\Kiddo.Web\Controllers\ProfileController.cs:line 38
   at lambda_method123(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeInnerFilterAsync>g__Awaited|13_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeFilterPipelineAsync>g__Awaited|20_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Awaited|17_0(ResourceInvoker invoker, Task task, IDisposable scope)
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
   at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
   at Swashbuckle.AspNetCore.SwaggerUI.SwaggerUIMiddleware.Invoke(HttpContext httpContext)
   at Swashbuckle.AspNetCore.Swagger.SwaggerMiddleware.Invoke(HttpContext httpContext, ISwaggerProvider swaggerProvider)
   at Serilog.AspNetCore.RequestLoggingMiddleware.Invoke(HttpContext httpContext)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)
```

**Comment by oledid** [comment]
Ok, the microsoft graph api is crashing at `Kiddo.Web.Models.ProfileModel.EnsureProfile()`, in the actual "get me"-request. I might be able to look at it some more tomorrow.

**Comment by kyledevans** [comment]
Success!  Calling `.WithAuthenticationScheme("AzureAd")` now works as expected.  Thank you @oledid, you've been extremely helpful in this!

Here's the actual call to the Graph API that now succeeds:
```
User? graphUser = await GraphClient.Me.Request().WithAuthenticationScheme("AzureAd").GetAsync().ConfigureAwait(false);
```
The authentication setup in Program.cs is still what I posted in https://github.com/AzureAD/microsoft-identity-web/issues/1624#issuecomment-1079598593.

It would be nice if the microsoft-identity-web library would automatically handle this sort of thing, but at least there's a clean workaround for now.  Perhaps implementing this sort of behavior in a way that covers every possible developer's usecase is an impossible task.  After looking at the internal logic the library uses to determine auth schemes it's very clear that this is not a straightforward problem to solve.

**Comment by oledid** [comment]
Great! Glad I could help.

`microsoft-identity-web` should probably check if `Instance` is null before calling `Instance = Instance.TrimEnd('/') + "/"` inside `PrepareAuthorityInstanceForMsal()`, but catching that the reason that `Instance` is null usually seems to be not setting correct AuthenticationScheme somewhere would be more helpful. But I'm not 100% sure that authenticationScheme is the problem in all cases where the `NullReferenceException` happens. But it was in my case and in @kyledevans' case.

**Comment by jmprieur** [comment]
https://github.com/AzureAD/microsoft-identity-web/pull/1672

**Comment by jennyf19** [comment]
Released in [1.24.0](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.24.0).
